### PR TITLE
feat(SemVer): Add Semantic Versioning docs to repo and link it from i…

### PR DESCRIPTION
**Pull request by ryekerjh**
…ntro page

This PR introduces our standards for SemVer (newly adopted as of the last S&P two weeks ago) and also links to it from the main repo README.md.

Please review the language, layout, and details, and test the links if you could. Let me know if I can improve anything.

### CODE REVIEW

File: git-and-github/code-versioning.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+## What is SemVer?
+
+Ever used NPM before? How about Homebrew, or RubyGems? Each of those and most other package management systems need a way of keeping track of what version their public codebase is on, what versions they have had before, and what version they are heading toward.
+
+The three package management systems I mentioned above all use a system of organization called [**Semantic Versioning**](https://semver.org/) to maintain uniform control of their code and how it is recieved by thier users. In short, versioning helps code maintainers and users of open source code identify several aspects of the code in question.
+
+## SemVer Basics:
+  * Always have an up-to-date documented API (in our case, README.md for our projects or Wiki)
+  * Increment your versions based on a **Major.Minor.Patch** format where:
+    * **Major** changes are breaking, non-backwards compatable changes to your code base.
```
Comment (carlosvargas) [inline git-and-github/code-versioning.md]:
```suggestion
    * **Major** changes are breaking, non-backwards compatible changes to your code base.
```

File: git-and-github/code-versioning.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    * **Minor** changes are new features or functionality that ius still backwards-compatable
```
Comment (carlosvargas) [inline git-and-github/code-versioning.md]:
```suggestion
    * **Minor** changes are new features or functionality that is still backwards-compatible
```

File: git-and-github/code-versioning.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  * Initial development (so most of our lives) should be treated as 0.x.x so as to indicate that there is no release as yet.
+    * Generally, any software with a major version of 0 should be treated as unstable.
+
+## Why should Shift3 care?
+
+This should be self-evident: more maintainable code makes our jobs easier! Semantic Versioning does the following things  
+  * Provides markers for points in our development cycle such as releases, patches, hotfixes, etc.
+  * Keeps in-development projects organized by allowing for pre-releases (ie. 0.1.0, 0.2.0) that can be mapped to specific milestones
+  * Allows for quicker set-up time for new developers being added to your project
+  * Help us avoid introducing major breaking changes unintentionally for enforcing the major/minor/patch methodology.
```
Comment (carlosvargas) [inline git-and-github/code-versioning.md]:
```suggestion
  * Help us avoid introducing major breaking changes unintentionally by enforcing the major/minor/patch methodology.  
```

File: git-and-github/code-versioning.md (hunk 1 above)
Comment (ryekerjh) [reply git-and-github/code-versioning.md]: Gooood catch!

**Comment by ryekerjh** [comment]
@carlosvargas Gooooood suggestions! I implemented all of them. Thanks for the catch!

**Comment by DropsOfSerenity** [comment]
@ryekerjh 

I'm unclear how SemVer is useful on a project level.

Rule 1 of the SemVer specification is:
image

Since many projects don't export a public api like a library would, It seems unclear when we would increment version numbers and what benefit we are getting.

**Comment by zbyte64** [comment]
@DropsOfSerenity For private projects I would argue that standard APIs used internally by the developers should be treated as a "public" API. For example, if a project changes how the developer is expected to retrieve data or depends on a framework that requires backward breaking changes, that should require a major version increment.

**Comment by ryekerjh** [comment]
@DropsOfSerenity As I mentioned in the meeting when I presented this, we aren't necessarily adopting the strict methodology of SemVer per se, but rather taking the advantages we can get out of it (i.e. versioning that's tied to milestones and releases of a project. Especially with v2 and v3 type projects coming in more and more, I think that taking advantage of that alone is going to be beneficial. I also agree with @zbyte64 in that we should treat each project as public API. 

I can enumerate these benefits in this markdown sheet to make them clearer.

**Comment by ryekerjh** [comment]
@DropsOfSerenity After our talk in the office, are you still feeling like this is unnecessary?

**Comment by ryekerjh** [comment]
@DropsOfSerenity I changed this pretty significantly to move away from strict adherence to SemVer specifically and to push adoption of Git Tagging and Releases more strictly. LMK what you think.

@mwallert @carlosvargas Can you take another once-over look at this?

File: git-and-github/code-versioning.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# Git Tagging & Releases
+
+## What is Git Tagging?
+
+  * [Tagging](https://git-scm.com/book/en/v2/Git-Basics-Tagging) is a simple, easy way to keep track of your code at a certain point in time. Some instances where this becomes useful are:
+    * Client-facing deliverables such as milestones, UAT, or production launch/release
+    * Internal-facing code at a point that you or your team wants to demarcate as important
```
Comment (carlosvargas) [inline git-and-github/code-versioning.md]: > demarcate

You don't gotta get all fancy words on us lol

**Comment by stephengtuggy** [comment]
I would suggest that we use a "SemVer-like" versioning system, where we don't adhere strictly to SemVer for UI projects and such (because that isn't possible) but we do use something that follows SemVer as much as possible. For example, in the Ultra-Gro project I am working on, we are in Phase 2. So for the UAT phase, I used version numbers v2.0.0-beta.1, v2.0.0-beta.2, etc.

On FUSD, when I change .NET Framework versions from 4.5.1 to 4.5.2 or similar, I change the minor version number of the project as well. Maybe I should change the major version number instead; I dunno. I have also changed the minor version number of the project when introducing TypeScript, or separating out dependency projects between Selection Portal and Summer School -- stuff like that.

What do you guys think?

**Comment by ryekerjh** [comment]
@stephengtuggy I see the use in that, but I agree wiht @DropsOfSerenity that the versioning does not need to adhere strictly to numerical tags. I think @carlosvargas has a good point in that we could use human-readable names instead of numbers that get incremented. 

What if we use some naming convention such as `pre-release-xx` for everything before deployment, then `prod-xx.yy` (replacing `xx` and `yy` with meaningful sub-types such as we do with our Karma Commits like `patch-ui-fix` or `feat-add-uploads`) for everything after production launch. 

What do you guys think @carlosvargas @stephengtuggy @DropsOfSerenity ?

**Comment by DropsOfSerenity** [comment]
@ryekerjh I like this a lot more. Perhaps in another standard we can talk about using something like git flow to even further refine our commit process.

**Comment by ryekerjh** [comment]
> @ryekerjh I like this a lot more. Perhaps in another standard we can talk about using something like git flow to even further refine our commit process.

soooo......approved? @DropsOfSerenity

File: git-and-github/code-versioning.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    * Versioning of your code (such as following SemVer or other release formats)
+  * Because we are not following a strict rule or methodology, each of our teams can decide for themselves which method of tag naming they want to follow. You could go strict and version your code based on milestones or releases, or you could use human-readable tag names. In the end, the power is _yours_!
```
Comment (ggoforth) [inline git-and-github/code-versioning.md]: I would challenge us to find and commit to one way to do this so we aren't reinventing the wheel here with each project.

Comment (ggoforth) [reply git-and-github/code-versioning.md]: That being said, as a whole I like this version, and think it's a good place to start.

**Comment by coreyshuman** [comment]
@ryekerjh @carlosvargas @ggoforth @DropsOfSerenity @stephengtuggy @chawkins-shift3 @Karvel 
I'm with Greg, this needs to be standardized across Shift3, not different per-project. I suggest we stick with SemVer style, we've used it across a dozen or more Shift3 projects and haven't run into any issues that I'm aware of. 
- It is much clearer to follow version progression using numbers than text/number combo. 
- Our SemVer values internally can and should match those which are included in release version numbers on the app stores. This makes it WAY easier to track and manage and troubleshoot.
- SemVer is also easy to communicate with clients when include in release notes.

Yeah, I'm really leaning towards if it's not broke don't fix it in this case.

**Comment by DropsOfSerenity** [comment]
I understand the desire to standardize, but I still don't see how SemVer on an application isn't anything but arbitrary or marketing decided anyway. If you keep a clean git history and follow something like git flow, your git history is your release history already, no versioning necessary.

I'm worried about the extra overhead and arbitrariness of SemVer on application projects where the concept of a "public api" makes no sense.

**Comment by DropsOfSerenity** [comment]
I do agree that some loose version of SemVer for applications that appear on stores is necessary though (since they do request a version number). But again, I'm a bit confused as to how we can make that a consistent thing, as opposed to something decided per project and per marketing decisions anyway. What does a minor, major, patch, version mean int he context of something that doesn't expose an api. What does it mean to the user?

**Comment by stephengtuggy** [comment]
@DropsOfSerenity Extra overhead? Seriously? It takes like two minutes to tag a release on GitHub. I don't understand what you're getting at here.

Arbitrariness? How about bumping the major version number when the underlying technology stack changes significantly, or when user-visible behavior changes in a backwards-incompatible way? When we add features without breaking backward compatibility in terms of functionality, then we bump the minor version number. And for bugfixes, security patches, etc., we bump the patch version number. Make sense?

(To clarify, this would be a departure from how we -- or at least I -- currently do versioning. No more "Phase 2 = v2.x".)

I genuinely do not understand how one can get by without actual version numbers / tags of some kind. Gitflow is great, and so is keeping a clean git history, but these techniques do nothing to address the need for concrete, addressable point releases, such as for support purposes. (E.g., the user reports that "The password reset button is broken in v2.3.0.") But maybe I misunderstood what you were saying.

SemVer-ish version numbers are more easily sortable. That makes it easier for users to see them in order, and to understand which ones came first and which ones came later. At least IMO.

You do make a good point, though, that we can't strictly adhere to the SemVer standard, because it requires you to have a public **API**. But I think we can use a "SemVer-ish" style, like the one I have outlined, and like @zbyte64 described.

I respect all of your opinions. I hope we can get this hammered out and come to an agreement soon.

**Comment by stephengtuggy** [comment]
Oh, perhaps one other thing needs to be clarified here: We aren't going to version / tag every commit, every branch, or even every pull request, right? Just stuff that goes into QA testing, into UAT, into production, or through one of the app stores?

**Comment by ryekerjh** [comment]
@stephengtuggy you are correct, it would _not_ be every commit or successful merge into the base branch. It would be only the release versions or milestone markers (if in pre production.) I echo your hope that we can get this ironed out and resolved soon!

**Comment by Karvel** [comment]
Let's not lose sight of the fact that currently there is no Shift3 standard for versioning. A SemVer-like solution fixes far more problems than it creates. We will be doomed to failure in all standards if we insist on holding out for perfection.

I support this proposal. If we find there are aspects that we need to adjust, we will adjust them to meet our real use cases, not hypotheticals.

**Comment by coreyshuman** [comment]
We are getting into the weeds here. I appreciate everyone's feedback, but I'm making a final call here so we can move forward:
- We WILL be using SemVer numbering for tags, in release notes, app store versioning, documentation, and in software and API version numbers. Said numbering should match across all of those components in each project.
- Note this does NOT meaning we are strictly following the Semantic Versioning Specification. Instead we are using it as a starting point for our own defined standards.

@ryekerjh I will send you some specific feedback on changes I'd like to see, and then we'll get this merged in.

**Comment by ryekerjh** [comment]
Sounds good @coreyshuman

File: git-and-github/code-versioning.md (hunk 5 above)
Comment (ryekerjh) [reply git-and-github/code-versioning.md]: @ryekerjh `major.minor.patch`

**Comment by DropsOfSerenity** [comment]
Sounds good, I'm good with getting this merged in, as it's better than what we have been doing so far. I'm sure we will be able to refine it over time, as to what it means for a backend repo, vs an app store application.

Merge away, this ones been open too long :smile: :+1:

**Comment by ryekerjh** [comment]
@coreyshuman I fixed this up and made those changes you requested. Merge?

**Pull request merged**